### PR TITLE
cxxrtl: Add unit tests for runtime

### PR DESCRIPTION
**Pull request by merryhime**
Proposed testing strategy for the cxxrtl runtime.
This PR also fixes a few bugs found with randomized unit testing.

Use of `patchelf` is requried in order to override `libyosys.so` SONAME. The SONAME provides an absolute path into system library directory, which is not what we want here as we would like to test the currently built library and not the installed system library.

### CODE REVIEW

**Review by whitequark — CHANGES_REQUESTED** [review]
I would rather not use gtest. I think the gtest stuff in `test/units` is kind of cursed and not really used, and having to use `patchelf` makes me sad. I think I would personally be happy with just a series of `assert` s

**Comment by merryhime** [comment]
Revisit.

**Comment by povik** [comment]
Ah, adding the gtest stuff to CI is the subject of the last commit

File: backends/cxxrtl/runtime/cxxrtl/cxxrtl.h [hunk 1]
```diff
@@ -444,12 +444,13 @@ struct value : public expr_base<value<Bits>> {
...
 		}
 		if (Signed && is_neg()) {
-			size_t top_chunk_idx  = (Bits - shift_bits) / chunk::bits;
-			size_t top_chunk_bits = (Bits - shift_bits) % chunk::bits;
+			size_t top_chunk_idx  = amount.data[0] > Bits ? 0 : (Bits - amount.data[0]) / chunk::bits;
+			size_t top_chunk_bits = amount.data[0] > Bits ? 0 : (Bits - amount.data[0]) % chunk::bits;
 			for (size_t n = top_chunk_idx + 1; n < chunks; n++)
 				result.data[n] = chunk::mask;
-			if (shift_bits != 0)
+			if (amount.data[0] != 0)
```
Comment (povik) [inline backends/cxxrtl/runtime/cxxrtl/cxxrtl.h:451]: Maybe we could make this check `top_chunk_idx < chunks`

**Comment by povik** [comment]
I reviewed the fixes, they all look good, I will cherry-pick them into a PR of their own, so that they are not blocked on resolving what the testing should be.

**Pull request closed without merging**